proposals/TP1-xds-transport-next.md

```diff
@@ -0,0 +1,864 @@
+4. Server responds with resources contents for
+   `xdstp://some-authority/envoy.config.listener.v3.Listener/bar `and
+   `xdstp://some-authority/envoy.config.listener.v3.Listener/baz`.
+
+This involves two round-trips. Inlining of resources is supported, which can reduce RTT, i.e.
+
+1. Client requests `xdstp://some-authority/envoy.config.listener.v3.ListenerCollection/foo.`
+2. Server responds with a collection of literal resources inlined in `foo`:
+   `[inline_entry: xdstp://some-authority/envoy.config.listener.v3.Listener/bar,
+   inline_entry: xdstp://some-authority/envoy.config.listener.v3.Listener/baz].`
+
+Note that in all cases, the context parameters presented in the original request
+must be present in the response. For example, a request for
+`xdstp://some-authority/envoy.config.listener.v3.ListenerCollection/foo?some=thing`, must
+be responded to with a resource named
+`xdstp://some-authority/envoy.config.listener.v3.ListenerCollection/foo?some=thing`.
+However, individual resources referenced by locator inside the returned list
+collection do not need to have the collection-level context parameters included
+in their naming.
+
+#### Glob
+
+Having explicit collection resources allows delta xDS to provide delta updates
+on individual resources but the resource list itself can become a bottleneck for
+frequently updated large collections of resources, e.g. with 1M endpoints updating
+every 10s.  For this situation, we introduce *glob
+collections*, which allow the `<T>Collection` directory to be elided when
+resources in a collection share a directory structure and a `/*` component is appended to
+the resource path. Continuing the previous example, this now looks like:
+
+1. Client requests
+   `xdstp://some-authority/envoy.config.listener.v3.Listener/foo/*`.
+2. Server responds with resources
+   [`xdstp://some-authority/envoy.config.listener.v3.Listener/foo/bar,
+   xdstp://some-authority/envoy.config.listener.v3.Listener/foo/baz]`.
+
+Note that there is no explicit directory sent enumerating `{bar, baz}`. Rather,
+glob collections provide delta updates directly on directory resources.
+
+Globs can exist in arbitrary path locations, e.g.
+`xdstp://some-authority/envoy.config.listener.v3.Listener/some/longer/path/*`.
+Multiple globs may be subscribed to in a `DeltaDiscoveryRequest`.
+
+Since each resource returned is subject to independent update via delta xDS and
+there is no explicit collection directory to update, glob collections are highly
+scalable. For example, consider a collection of 10k resources and the addition
+of a single resource. Glob collections will send the single additional resource
+(due to the operation of delta xDS). List collections requires that the
+collection directory (now containing 10001 resource references) to be sent from
+server to client, as well as the additional new resource.
+
+A server may reply with a sentinel `Resource` with no value populated and
```

Comment:
I see how this can work for many examples.

It might be my misinterpretation of `removed_resources`, but I might have a use-case that this doesn't cover:
1. Client's bootstrap is set to fetch a glob collection of clusters:
```
dynamic_resources:
  cds_resources_locator:
    name: xdstp://some-authority/envoy.config.clusters.v3.Cluster/my-clusters/*
  cds_config:
    authorities:
    - name: some-authority
```
2. At the moment the (due to some transient error) the management server has no clusters configured, and so it responds with `removed_resources:["xdstp://some-authority/envoy.config.clusters.v3.Cluster/my-clusters/*"]`.
3. The client receives the response and is no longer registered to that glob list.

I think that the sentinel allows to keep the registration alive.
WDYT?

Reply:
Adi, the server telling the client that the resource does not exist does not cause the client to unsubscribe from the resource.  Only the client controls what resources it is subscribed to; the server cannot alter that list.  The client will still be subscribed, and if the resource is created later, the server will send an update containing the resource.

You can think of "resource does not exist" as just a special sentinel value of a given resource.  While the client is subscribed to a resource, the server sends updates whenever the resource changes from one value to another, and "does not exist" is just another possible value for the resource.  Whenever the resource's value changes from some other value to "does not exist" or from "does not exist" to some other value, the server will send an update, just like it would on any other value change.

Reply:
Updated to this scheme, I don't see a big deal either way.